src/Lecture2.hs

```diff
@@ -242,11 +242,17 @@ The algorithm of merge sort is the following:
 mergeSort :: [Int] -> [Int]
 mergeSort [] = []
 mergeSort [x] = [x]
-mergeSort xs = let
-    halfLen = length xs `quot` 2
-    (firstHalf, secondHalf) = splitAt halfLen xs
-    in
-    merge (mergeSort firstHalf) (mergeSort secondHalf)
+mergeSort xss = (mergePairs . pairify) xss
+    where
+    pairify :: [Int] -> [[Int]]
+    pairify [] = []
+    pairify [x] = [[x]]
+    pairify (x:y:zs) = merge [x] [y] : pairify zs
+
+    mergePairs :: [[Int]] -> [Int]
+    mergePairs [] = []
+    mergePairs [xs] = xs
+    mergePairs (xs:ys:zss) = mergePairs (merge xs ys : zss)
```

Comment:
That's a clever and correct solution 😅 
Unfortunately, it's not optimal asymptotically. Your solution has complexity `O(n ^ 2)` while optimal merge sort has `O(n log n)`. But the standard `sort` function in Haskell implements an idea similar to yours and improves it a bit to achieve even better performance 🐎

This happens because you merge the first list with the second one. And in your case you merge two lists of size 1 and get size 2, then you merge a list of size 1 with this 2-elements lists and you get a list of size 3, then you merge this 3-elements list with a 1-size list again, and so on.

Fortunately, it's quite easy to fix in this case 👍🏻 You just need to recursively call `mergePairs` on the `zss` to halve the list size by two on each iteration.

You can move functions around and simplify this a bit to make the code look nicer but that's a general idea 🙂